### PR TITLE
Implementation for .isBetween method

### PR DESCRIPTION
**Pull request by sabbbin**

### CODE REVIEW

File: packages/bikram-sambat/README.md [hunk 1]
```diff
@@ -371,6 +371,51 @@ Return true if provided BikramSambat object is after the other supplied BikramSa
 | day   | Day         |
 | week  | Week        |
 
+#### .isBetween(startDate,endDate, unit, include) -> boolean
+
+Return true if provided BikramSambat object is between the other supplied BikramSambat date.
```
Comment (Nakarmi23) [inline packages/bikram-sambat/README.md]:
```suggestion
Return true if provided BikramSambat object is between the other supplied BikramSambat dates.
```

File: packages/bikram-sambat/README.md [hunk 2]
```diff
@@ -371,6 +371,51 @@ Return true if provided BikramSambat object is after the other supplied BikramSa
...
+
+- Parameters:
+
+  - `startDate` (BikramSambat): The BikramSambat object to compare against.
+  - `endDateDate` (BikramSambat): The BikramSambat object to compare against.
```
Comment (Nakarmi23) [inline packages/bikram-sambat/README.md]:
```suggestion
  - `startDate` (BikramSambat | Date): The BikramSambat object to compare against.
  - `endDateDate` (BikramSambat | Date): The BikramSambat object to compare against.
```

File: packages/bikram-sambat/README.md [hunk 3]
```diff
@@ -371,6 +371,51 @@ Return true if provided BikramSambat object is after the other supplied BikramSa
...
+  - `unit` (string): Optional. Unit of comparison. Defaults to `day`.
+  - `includes` (string): Optional. Unit of comparison. Defaults to `()`.
```
Comment (Nakarmi23) [inline packages/bikram-sambat/README.md]: You can use ChatGPT to help you with documentation

File: packages/bikram-sambat/README.md [hunk 4]
```diff
@@ -371,6 +371,51 @@ Return true if provided BikramSambat object is after the other supplied BikramSa
...
+| day   | Day         |
+
+##### List of all available includes
+
+| Unit | Description                           |
+| ---- | ------------------------------------- |
+| ()   | include startDate and endDate         |
+| []   | exclude startDate and endDate         |
+| [)   | exclude startDate and include endDate |
+| (]   | include startDate and exclude endDate |
```
Comment (Nakarmi23) [inline packages/bikram-sambat/README.md]: Write a more clear documentation. You can take help of ChatGPT

File: packages/bikram-sambat/src/index.ts [hunk 5]
```diff
@@ -363,4 +366,40 @@ export default class BikramSambat implements BikramSambatProps {
 
     return dayjs(this.adDate).isAfter(date, unit);
   }
+  /**
+   * Checks if the current BikramSambat date is after the given date.
+   * @param {BikramSambat | Date} startDate - The startdate to compare if  it is before given date.
+   * @param {BikramSambat | Date} endDate - The enddate to compare if it is  after the given date .
+   * @param {ManipulateType} [unit='day'] - The unit of time to compare. Defaults to 'day'.
+   * @param {IsBetweenIncludeExcludeType} [include='()'] - The include of time  compare. Defaults to '()'.
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts]: Use better name for IsBetweenIncludeExcludeType type

Comment (Nakarmi23) [reply packages/bikram-sambat/src/index.ts]: also add supported include tokens

File: packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import type { IsBetweenIncludeExcludeType } from '..';
+
+export const shouldIncludeStartEndDates = (
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts]:
```suggestion
export const parseIncludeExcludeToken = (
```

**Review by Nakarmi23 — CHANGES_REQUESTED** [review]

File: packages/bikram-sambat/src/index.ts [hunk 7]
```diff
@@ -18,6 +19,8 @@ interface BikramSambatProps {
 
 export type UnitType = 'day' | 'month' | 'year' | 'week';
 
+export type getBoundaryInclusionType = '()' | '[)' | '(]' | '[]';
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts]:
```suggestion
export type BoundaryInclusionType = '()' | '[)' | '(]' | '[]';
```

File: packages/bikram-sambat/src/index.ts [hunk 8]
```diff
@@ -363,4 +366,46 @@ export default class BikramSambat implements BikramSambatProps {
 
     return dayjs(this.adDate).isAfter(date, unit);
   }
+  /**
+   * Checks if the current BikramSambat date is after the given date.
+   *  * Formats the BikramSambat date according to the specified format string.
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts]:
```suggestion
```

File: packages/bikram-sambat/src/index.ts [hunk 9]
```diff
@@ -363,4 +366,46 @@ export default class BikramSambat implements BikramSambatProps {
 
     return dayjs(this.adDate).isAfter(date, unit);
   }
+  /**
+   * Checks if the current BikramSambat date is after the given date.
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts]:
```suggestion
   * Checks if the current BikramSambat date is between the given dates. 
```
Add some infor about boundary inclusion

**Review by Nakarmi23 — CHANGES_REQUESTED** [review]

File: packages/bikram-sambat/src/index.ts [hunk 10]
```diff
@@ -363,4 +366,47 @@ export default class BikramSambat implements BikramSambatProps {
...
+   * - '(]': Exclude startDate, include endDate
+   * - '[)': Include startDate, exclude endDate
+   *
+   * @param {BikramSambat | Date} startDate - The start date to compare against.
+   * @param {BikramSambat | Date} endDate - The end date to compare against.
+   * @param {ManipulateType} [unit='day'] - The unit of time for comparison. Defaults to 'day'.
+   * @param {boundaryInclusionType} [boundaryInclusion='()'] - The boundary inclusion type for comparison. Defaults to '()'.
+   * @returns {boolean} True if the date is between startDate and endDate according to the specified boundaryInclusion, false otherwise.
+   */
+
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts:384]:
```suggestion
```

File: packages/bikram-sambat/src/index.ts [hunk 11]
```diff
@@ -18,6 +19,8 @@ interface BikramSambatProps {
 
 export type UnitType = 'day' | 'month' | 'year' | 'week';
 
+export type boundaryInclusionType = '()' | '[)' | '(]' | '[]';
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts]:
```suggestion
export type BoundaryInclusionType = '()' | '[)' | '(]' | '[]';
```

File: packages/bikram-sambat/README.md [hunk 12]
```diff
@@ -346,7 +346,7 @@ Return true if provided BikramSambat object is before the other supplied BikramS
 
 #### .isAfter(date, unit) -> boolean
 
-Return true if provided BikramSambat object is after the other supplied BikramSambat date.
+Return true if provided BikramSambat object is after the other supplied BikramSambat dates.
```
Comment (Nakarmi23) [inline packages/bikram-sambat/README.md]:
```suggestion
Return true if provided BikramSambat object is after the other supplied BikramSambat date.
```

**Review by Nakarmi23 — CHANGES_REQUESTED** [review]

File: packages/bikram-sambat/src/index.ts [hunk 13]
```diff
@@ -363,4 +366,47 @@ export default class BikramSambat implements BikramSambatProps {
...
+   * @param {ManipulateType} [unit='day'] - The unit of time for comparison. Defaults to 'day'.
+   * @param {BoundaryInclusionType} [boundaryInclusion='()'] - The boundary inclusion type for comparison. Defaults to '()'.
+   * @returns {boolean} True if the date is between startDate and endDate according to the specified boundaryInclusion, false otherwise.
+   */
+
+  isBetween(
+    startDate: BikramSambat | Date,
+    endDate: BikramSambat | Date,
+    unit: ManipulateType = 'day',
+    boundaryInclusion: BoundaryInclusionType = '()'
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts:389]: add robust validation for boundary inclusion token using pattern matching and also add test for it

File: packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+): [number, number] => {
+  switch (val) {
+    case '()':
+      return [1, 1];
+    case '(]':
+      return [1, 0];
+    case '[)':
+      return [0, 1];
+    case '[]':
+      return [0, 0];
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts:14]: in default case throw error

File: packages/bikram-sambat/src/bikram-sambat.test.ts [hunk 15]
```diff
@@ -189,4 +189,65 @@ describe('Bikram Sambat', () => {
 
     expect(() => date.isBefore('2024-03-03' as never)).toThrowError();
   });
+
+  test('Date Between: isBetween', () => {
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/bikram-sambat.test.ts:193]: add test if user enters 'abc' or other this in case of '()' or others

**Review by Nakarmi23 — CHANGES_REQUESTED** [review]

File: packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+    default:
+      throw new Error('boundary inclusion does not match');
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts]:
```suggestion
      throw new Error(`Invalid boundary inclusion: ${value}`);
```

File: packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts [hunk 17]
```diff
@@ -0,0 +1,18 @@
+import type { BoundaryInclusionType } from '..';
+
+export const getBoundaryInclusion = (
+  val: BoundaryInclusionType
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/utilities/get-bs-data-isbetween.ts]:
```suggestion
  value: BoundaryInclusionType
```

File: packages/bikram-sambat/src/index.ts [hunk 18]
```diff
@@ -363,4 +366,57 @@ export default class BikramSambat implements BikramSambatProps {
...
+  ): boolean {
+    if (startDate instanceof BikramSambat) startDate = startDate.adDate;
+    else if (!(startDate instanceof Date))
+      throw new Error('Invalid compare value');
+    if (endDate instanceof BikramSambat) endDate = endDate.adDate;
+    else if (!(endDate instanceof Date))
+      throw new Error('Invalid compare value');
+
+    const boundaryInclusionpattern = /^(\(\)|\[\]|\(\]|\[\))$/;
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts]:
```suggestion
    const boundaryInclusionPattern = /^[\(\[][\)\]]$/;
```

File: packages/bikram-sambat/src/index.ts [hunk 19]
```diff
@@ -363,4 +366,57 @@ export default class BikramSambat implements BikramSambatProps {
...
+    const unitPattern = /^(day|year|month)$/;
+
+    if (
+      !boundaryInclusionpattern.test(boundaryInclusion) ||
+      !unitPattern.test(unit)
+    ) {
+      throw new Error('Invalid Bounary inclusion pattern');
+    }
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/index.ts]: for now dont put unit validation, will include in next patch update

Comment (Nakarmi23) [reply packages/bikram-sambat/src/index.ts]:
```suggestion
      !boundaryInclusionpattern.test(boundaryInclusion) ||
      !unitPattern.test(unit)
    ) {
      throw new Error('Invalid bounary inclusion pattern');
    }
```

File: packages/bikram-sambat/src/bikram-sambat.test.ts [hunk 20]
```diff
@@ -189,4 +189,76 @@ describe('Bikram Sambat', () => {
...
+    expect(yendDatec1.isBetween(ybeginDate, ydate, 'year')).toBeTruthy();
+    expect(() =>
+      yendDatec1.isBetween('' as never, ydate, 'year')
+    ).toThrowError();
+    expect(() =>
+      yendDatec1.isBetween(ybeginDate, '' as never, 'year')
+    ).toThrowError();
+
+    expect(() =>
+      yendDatec1.isBetween('' as never, ydate, 'year', ')(' as never)
```
Comment (Nakarmi23) [inline packages/bikram-sambat/src/bikram-sambat.test.ts:254]: also test for abc in boundary inclusion

**Review by Nakarmi23 — CHANGES_REQUESTED** [review]

**Review by Nakarmi23 — APPROVED** [review]
